feat: Extend MFE Config API for `frontend-app-catalog` [FC-0086]

Serj-N opened the pull request:
**Summary**
This PR extends the MFE Config API in order to return MFE configuration values using the following hierarchy, in the order from most specific to least specific:

1. site configuration MFE_CONFIG_OVERRIDES
2. settings MFE_CONFIG_OVERRIDES
3. site configuration MFE_CONFIG
4. settings MFE_CONFIG
5. site configuration plain value
6. settings plain value

The following fields have been added to the configuration being returned:

- ENABLE_COURSE_SORTING_BY_START_DATE
- HOMEPAGE_PROMO_VIDEO_YOUTUBE_ID
- HOMEPAGE_COURSE_MAX
- COURSE_ABOUT_TWITTER_ACCOUNT
- NON_BROWSABLE_COURSES
- ENABLE_COURSE_DISCOVERY

**Usage**
Usage remains exactly as before:
Get common config:
`GET /api/mfe_config/v1`

Get app config (common + mfe-specific overrides):
`GET /api/mfe_config/v1?mfe=name_of_mfe`

**Example of a GET response**
(abbreviated)
```
{
  "ENABLE_COURSE_SORTING_BY_START_DATE": true,
  "HOMEPAGE_PROMO_VIDEO_YOUTUBE_ID": "your-youtube-id",
  "HOMEPAGE_COURSE_MAX": 9,
  "BASE_URL": "apps.local.openedx.io",
  "CSRF_TOKEN_API_PATH": "/csrf/api/v1/token",
  "CREDENTIALS_BASE_URL": "",
  ...
}
```

**Dependent PRs:**
[feat: [FC-86] Home page banner section](https://github.com/openedx/frontend-app-catalog/pull/7)
[feat: [FC-86] Added Home page courses list](https://github.com/raccoongang/frontend-app-catalog/pull/16)
[feat: [FC-86] Created Intro section for Course About page](https://github.com/raccoongang/frontend-app-catalog/pull/19)
[feat: [FC-86] Added Course About sidebar](https://github.com/raccoongang/frontend-app-catalog/pull/20)
[feat: [FC-86] Added course overview and updated sidebar](https://github.com/raccoongang/frontend-app-catalog/pull/21)
[feat: [FC-86] Updated Header links](https://github.com/raccoongang/frontend-app-catalog/pull/18)
[feat: [FC-86] Added DataTable for Catalog page](https://github.com/raccoongang/frontend-app-catalog/pull/22)
[feat: [FC-86] Added filtration for Catalog page](https://github.com/raccoongang/frontend-app-catalog/pull/23)
[feat: [FC-86] Added course search for DataTable](https://github.com/raccoongang/frontend-app-catalog/pull/24)

Serj-N committed:
feat: return response merged from base config, mfe config and mfe overrides

Serj-N committed:
test: adjust existing tests

Serj-N committed:
test: add new test methods

openedx-webhooks commented:
Thanks for the pull request, @Serj-N!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

Serj-N committed:
fix: add blank line

cmltaWt0 commented on lms/djangoapps/mfe_config_api/tests/test_views.py:
```diff
@@ -11,12 +11,25 @@
 
+# Based config values, used in tests to build a correct expected response
+default_base_config = {
+    'course_about_twitter_account': '@YourPlatformTwitterAccount',
```
Do we still use Twitter even if it is called X now? Just wondering, no actions required.

cmltaWt0 commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +69,54 @@ def get(self, request):
+                "ENABLE_COURSE_SORTING_BY_START_DATE",
+                settings.FEATURES["ENABLE_COURSE_SORTING_BY_START_DATE"]
+            ),
+            "homepage_overlay_html": configuration_helpers.get_value("homepage_overlay_html"),
```
This field should be removed from the list since we are going to implement it via a plugin slot.

Serj-N replied:
done

cmltaWt0 commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +69,54 @@ def get(self, request):
+                "course_about_twitter_account",
+                settings.PLATFORM_TWITTER_ACCOUNT
+            ),
+            "is_cosmetic_price_enabled": settings.FEATURES.get("ENABLE_COSMETIC_DISPLAY_PRICE"),
```
Need to clarify whether we need a MFe config here? I believe the preliminary decision was to allow to override from MFE_CONFIG.

brian-smith-tcril replied:
This is still marked as "Unknown" in the "Do we want this functionality?" column of the spreadsheet.

I left a comment there asking:
> Could you elaborate on this? I found https://github.com/openedx/edx-platform/blob/0bf88d225f563ad5a81205706a5ed9be91fe0794/lms/templates/courseware/course_about.html#L222-L228 and it's not clear to me how the "when there is no registration price" part of this fits in

Serj-N replied:
Removed as discussed in [the spreadsheet](https://docs.google.com/spreadsheets/d/17q5Kh7CSxe5JSNeNAA3VzeMBuD0WBTxNfDF7k2kmtbI/edit?gid=0#gid=0)

cmltaWt0 commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +69,54 @@ def get(self, request):
+                settings.PLATFORM_TWITTER_ACCOUNT
+            ),
+            "is_cosmetic_price_enabled": settings.FEATURES.get("ENABLE_COSMETIC_DISPLAY_PRICE"),
+            "courses_are_browsable": settings.FEATURES.get("COURSES_ARE_BROWSABLE")
```
Non clear decision yet but it looks like we are leaning toward the following fallback in frontend-app-learner-dashboard:
```
NON_BROWSABLE_COURSES fall back to !(not)COURSES_ARE_BROWSABLE
```

Serj-N replied:
Changed to `NON_BROWSABLE_COURSES`

cmltaWt0 commented on lms/djangoapps/mfe_config_api/tests/test_views.py:
```diff
@@ -136,7 +163,7 @@ def test_get_mfe_config_from_django_settings(self):
         response = self.client.get(self.mfe_config_api_url)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
-        self.assertEqual(response.json(), settings.MFE_CONFIG)
+        self.assertEqual(response.json(), settings.MFE_CONFIG | default_base_config)
```
Looks like it differs from this order:
```python
 # Merge the three configs in the order of precedence
merged_config = base_config | mfe_config | mfe_config_overrides
```

Could you explain why?

Serj-N replied:
This was appended to make the existing tests pass, and in those particular tests the order of precedence didn't matter much, considering the test values returned. But you are right, it is better to use the established order everywhere to avoid confusion. Thanks for noticing. Done.

Serj-N committed:
fix: remove homepage_overlay_html

Serj-N committed:
fix: typo in a comment

Serj-N committed:
test: change dict merge order to follow the correct hierarchy

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +69,53 @@ def get(self, request):
+            "show_homepage_promo_video": configuration_helpers.get_value(
+                "show_homepage_promo_video",
+                False
+            ),
```
I believe we can remove this.

From our conversations on the spreadsheet:
> @brian-smith-tcril: I added a "Yes (needs DEPR)" option to this dropdown. This is functionality we want to keep, but I believe the implementation change will require a DEPR.

> @PKulkoRaccoonGang: Am I understanding correctly that the approval will come after we pass the DEPR?

> @brian-smith-tcril: My understanding is that we should:
> * Not expose `show_homepage_promo_video` to the MFE
> * Ensure the functionality provided by the `show_homepage_promo_video` toggle (having a homepage without a promo video) still exists in the MFE. This will be covered by the suggested check for a valid `homepage_promo_video_youtube_id` value and a plugin slot.
> * During the release where site operators can choose to use the legacy pages or the MFE, we will have a large DEPR covering the removal of the legacy pages
> * We need to make sure to note this change in that DEPR

Serj-N replied:
Removed

PKulkoRaccoonGang commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +69,53 @@ def get(self, request):
+                "show_homepage_promo_video",
+                False
+            ),
+            "homepage_promo_video_youtube_id": configuration_helpers.get_value(
```
According to the case of other fields in MFE config it makes sense for the new ones to also be in the format upper case

Serj-N replied:
Changed it to uppercase (even though apparently the [official docs](https://docs.openedx.org/en/latest/site_ops/install_configure_run_guide/configuration/sites/configure_site.html) do not enforce this).
Done.

Serj-N committed:
fix: change response keys to uppercase

Serj-N committed:
fix: add enable_course_discovery

Serj-N committed:
fix: change COURSES_ARE_BROWSABLE to NON_BROWSABLE_COURSES

Serj-N committed:
fix: remove show_homepage_promo_video

sarina commented:
We've updated [the spreadsheet](https://docs.google.com/spreadsheets/d/17q5Kh7CSxe5JSNeNAA3VzeMBuD0WBTxNfDF7k2kmtbI/edit?gid=0#gid=0) and I think we're quite close to this. @Serj-N if you could work on fixing the tests I think we can take this out of draft when they pass and get @feanil 's review

feanil commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +69,50 @@ def get(self, request):
+            ),
+            "HOMEPAGE_PROMO_VIDEO_YOUTUBE_ID": configuration_helpers.get_value(
+                "homepage_promo_video_youtube_id",
+                "your-youtube-id"
```
Is this a default value that is being set in the code here?   Does it makes sense for this to be `None` by default? I'm not sure if it's important for this to be a valid string and my assumption is that if this is not some sort of null object, the consumer of the API might try to use it thinking it's valid and then have an unexpected failure because it wouldn't know that this is a string that is not a valid id.

Serj-N replied:
Good point, changed to `None`

feanil commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -31,6 +31,13 @@ class MFEConfigView(APIView):
     def get(self, request):
         """
         Return the MFE configuration, optionally including MFE-specific overrides.
```
```suggestion
        Return the MFE configuration, optionally including MFE-specific overrides.
        
        This configuration currently also pulls specific settings from site configuration or 
        django settings. This is a temporary change as a part of the migration of some legacy
        pages to MFEs.  This is a temporary compatibility layer which will eventually be deprecated.
        
        See [Link to DEPR ticket] for more details.
        
        The compatability means that settings from the legacy locations will continue to work but
        the settings listed below in the `_get_legacy_config` function should be added to the MFE
        config by operators.
```

Serj-N replied:
Docstring added with a todo (add link)

brian-smith-tcril replied:
@Serj-N would you be willing to create the new DEPR ticket for this and link to it here and in https://github.com/openedx/edx-platform/issues/36785?

brian-smith-tcril replied:
@Serj-N I landed this PR with the todo, and created an issue to track splitting up the DEPR and adding the link. https://github.com/openedx/edx-platform/issues/37210

feanil commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +69,50 @@ def get(self, request):
+        return JsonResponse(merged_config, status=status.HTTP_200_OK)
+
+    @staticmethod
+    def _get_base_config() -> dict:
```
```suggestion
    def _get_legacy_config() -> dict:
```

You'll need to update the calls as well but these are specifically a compatibility layer that is meant to be temporary, the naming should indicate that as should the docstring.

Serj-N replied:
Done

feanil reviewed (commented):
The general approach and code makes sense to me and I approve of it, I have one question where I'm looking for more information and a bit more documenting and ticketing that I'd like to see before this merges.

I think we should create the relevant DEPR for the dropping of the legacy settings from the MFE config and maybe fully from the settings as well if they're not used anywhere else.  That DEPR should be created and linked in the comments here so that it's easy to go find out why we're doing this in the future and come cleanup after ourselves.

I'm assuming the number of settings that are getting pulled here may change so I'm fine with creating the DEPR once that has all been settled but it should be done before we merge this code.

brian-smith-tcril commented:
> I think we should create the relevant DEPR for the dropping of the legacy settings from the MFE config and maybe fully from the settings as well if they're not used anywhere else. That DEPR should be created and linked in the comments here so that it's easy to go find out why we're doing this in the future and come cleanup after ourselves.

I know https://github.com/openedx/edx-platform/issues/36785 has been updated to include information about changes from this PR.

Serj-N committed:
fix: use None as default for promo video youtube id

Serj-N committed:
docs: expand docstrings, rename method/variables

Serj-N commented:
> We've updated [the spreadsheet](https://docs.google.com/spreadsheets/d/17q5Kh7CSxe5JSNeNAA3VzeMBuD0WBTxNfDF7k2kmtbI/edit?gid=0#gid=0) and I think we're quite close to this. @Serj-N if you could work on fixing the tests I think we can take this out of draft when they pass and get @feanil 's review

@sarina Now that I added new commits with the requested changes, the pipeline reran and all the tests pass. I went ahead and changed the PR status from draft to open, too.

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +72,50 @@ def get(self, request):
+        Return legacy configuration values available in either site configuration or django settings.
+        """
+        return {
+            "ENABLE_COURSE_SORTING_BY_START_DATE": configuration_helpers.get_value(
+                "ENABLE_COURSE_SORTING_BY_START_DATE",
+                settings.FEATURES["ENABLE_COURSE_SORTING_BY_START_DATE"]
+            ),
```
Perfect! This matches the decision in the spreadsheet to expose this as an MFE_CONFIG fallback!

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +72,50 @@ def get(self, request):
+                "ENABLE_COURSE_SORTING_BY_START_DATE",
+                settings.FEATURES["ENABLE_COURSE_SORTING_BY_START_DATE"]
+            ),
+            "HOMEPAGE_PROMO_VIDEO_YOUTUBE_ID": configuration_helpers.get_value(
+                "homepage_promo_video_youtube_id",
+                None
+            ),
```
Perfect! This matches the decision in the spreadsheet to expose this as an MFE_CONFIG fallback!

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +72,50 @@ def get(self, request):
+                "homepage_promo_video_youtube_id",
+                None
+            ),
+            "HOMEPAGE_COURSE_MAX": configuration_helpers.get_value(
+                "HOMEPAGE_COURSE_MAX",
+                settings.HOMEPAGE_COURSE_MAX
+            ),
```
Perfect! This matches the decision in the spreadsheet to expose this as an MFE_CONFIG fallback!

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +72,50 @@ def get(self, request):
+                "HOMEPAGE_COURSE_MAX",
+                settings.HOMEPAGE_COURSE_MAX
+            ),
+            "COURSE_ABOUT_TWITTER_ACCOUNT": configuration_helpers.get_value(
+                "course_about_twitter_account",
+                settings.PLATFORM_TWITTER_ACCOUNT
+            ),
```
Perfect! This matches the decision in the spreadsheet to expose this as an MFE_CONFIG fallback!

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +72,50 @@ def get(self, request):
+                "course_about_twitter_account",
+                settings.PLATFORM_TWITTER_ACCOUNT
+            ),
+            "IS_COSMETIC_PRICE_ENABLED": settings.FEATURES.get("ENABLE_COSMETIC_DISPLAY_PRICE"),
```
In the spreadsheet the decision is documented as:

> We should handle this in the same way as show_homepage_promo_video. If there isn't a price don't show it, and wrap the price in a plugin slot so people can remove it using FPF

so this should be removed.

```suggestion
```

Serj-N replied:
Done.
Thanks for a detailed breakdown!

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +72,50 @@ def get(self, request):
+                settings.PLATFORM_TWITTER_ACCOUNT
+            ),
+            "IS_COSMETIC_PRICE_ENABLED": settings.FEATURES.get("ENABLE_COSMETIC_DISPLAY_PRICE"),
+            "NON_BROWSABLE_COURSES": not settings.FEATURES.get("COURSES_ARE_BROWSABLE"),
```
Perfect! This matches the decision in the spreadsheet!

> Expose as MFE_CONFIG fallback. Update NON_BROWSABLE_COURSES to fall back to !COURSES_ARE_BROWSABLE.

brian-smith-tcril commented on lms/djangoapps/mfe_config_api/views.py:
```diff
@@ -60,12 +72,50 @@ def get(self, request):
+            ),
+            "IS_COSMETIC_PRICE_ENABLED": settings.FEATURES.get("ENABLE_COSMETIC_DISPLAY_PRICE"),
+            "NON_BROWSABLE_COURSES": not settings.FEATURES.get("COURSES_ARE_BROWSABLE"),
+            "ENABLE_COURSE_DISCOVERY": settings.FEATURES["ENABLE_COURSE_DISCOVERY"],
```
Perfect! This matches the decision in the spreadsheet to expose this as an MFE_CONFIG fallback!

brian-smith-tcril reviewed (commented):
This is looking great! I've gone through to check that the implementations match the decisions in the spreadsheet.

Here's an overview of what has been implemented at time of review:

| Status | Legacy Implementation |
| - | - |
| :white_check_mark: | settings.FEATURES["ENABLE_COURSE_SORTING_BY_START_DATE"]) |
| :outbox_tray: | homepage_overlay_html in Site Configuration or modifying index_overlay.html |
| :outbox_tray: | show_partners - in Site Configuration |
| :outbox_tray: | show_homepage_promo_video - in Site Configuration |
| :white_check_mark: | homepage_promo_video_youtube_id - in Site Configuration |
| :white_check_mark: | homepage_course_max - in Site Configuration or Django settings |
| :outbox_tray: | sidebar_html_enabled - WaffleSwitch "course_experience.enable_about_sidebar_html" |
| :outbox_tray: | course_about_show_social_links - in Site Configuration |
| :white_check_mark: | course_about_twitter_account - in Site Configuration or Django settings |
| :x: | is_cosmetic_price_enabled - settings.FEATURES['ENABLE_COSMETIC_DISPLAY_PRICE'] |
| :white_check_mark: | courses_are_browsable - settings.FEATURES['COURSES_ARE_BROWSABLE'] |
| :white_check_mark: | enable_course_discovery - settings.FEATURES['ENABLE_COURSE_DISCOVERY'] |

#### Status Key:
| Emoji | Meaning |
| - | - |
| :white_check_mark: | Implementation in this PR matches decision in spreadsheet |
| :outbox_tray: | Implementation **intentionally** not in this PR, matches decision in spreadsheet |
| :x: | Implementation in this PR **does not match** decision in spreadsheet

Serj-N committed:
fix: remove is_cosmetic_price_enabled field

brian-smith-tcril reviewed (approved):
Thank you so much for your patience throughout the review process on this one!

I'm very happy with how this turned out!

___

Here's an updated overview of what has been implemented at time of approval:

| Status | Legacy Implementation |
| - | - |
| :white_check_mark: | settings.FEATURES["ENABLE_COURSE_SORTING_BY_START_DATE"]) |
| :outbox_tray: | homepage_overlay_html in Site Configuration or modifying index_overlay.html |
| :outbox_tray: | show_partners - in Site Configuration |
| :outbox_tray: | show_homepage_promo_video - in Site Configuration |
| :white_check_mark: | homepage_promo_video_youtube_id - in Site Configuration |
| :white_check_mark: | homepage_course_max - in Site Configuration or Django settings |
| :outbox_tray: | sidebar_html_enabled - WaffleSwitch "course_experience.enable_about_sidebar_html" |
| :outbox_tray: | course_about_show_social_links - in Site Configuration |
| :white_check_mark: | course_about_twitter_account - in Site Configuration or Django settings |
| :outbox_tray: | is_cosmetic_price_enabled - settings.FEATURES['ENABLE_COSMETIC_DISPLAY_PRICE'] |
| :white_check_mark: | courses_are_browsable - settings.FEATURES['COURSES_ARE_BROWSABLE'] |
| :white_check_mark: | enable_course_discovery - settings.FEATURES['ENABLE_COURSE_DISCOVERY'] |

#### Status Key:
| Emoji | Meaning |
| - | - |
| :white_check_mark: | Implementation in this PR matches decision in spreadsheet |
| :outbox_tray: | Implementation **intentionally** not in this PR, matches decision in spreadsheet |

___

I also believe all of @feanil's comments have been addressed, so this should be good to merge!

brian-smith-tcril commented:
@cmltaWt0 have all of your change requests been addressed on this one?

cmltaWt0 commented:
> @cmltaWt0 have all of your change requests been addressed on this one?

Looking at it.
